workers/data_refinery_workers/processors/salmon.py

```diff
@@ -373,6 +396,58 @@ def _run_fastqc(job_context: Dict) -> Dict:
 
     return job_context
 
+def _get_salmon_completed_exp_dirs(job_context: Dict) -> List[str]:
+    """Return a list of directory names of experiments whose samples
+    have all been processed by `salmon quant` command.
+    """
+
+    sample = job_context['samples'][0]
+    experiments = ExperimentSampleAssociation.objects.filter(sample=sample)
+    salmon_completed_exp_dirs = []
+    salmon_cmd_str = 'salmon --no-version-check quant'
+    for experiment in experiments:
+        num_salmon_completed_samples = experiment.samples.filter(
+            results__command_executed__startswith=salmon_cmd_str).distinct().count()
+        if num_salmon_completed_samples == experiment.samples.count():
+            # Remove the last two parts from the path of job_context["original_files"]
+            # (which is "<experiment_accession_code>/raw/<filename>")
+            # to get the experiment directory name.
+            dir_tokens = job_context["original_files"][0].absolute_file_path.split('/')[:-2]
+            experiment_dir = '/'.join(dir_tokens)
+            salmon_completed_exp_dirs.add(experiment_dir)
+    return salmon_completed_exp_dirs
+
+
+def _tximport(exp_dir: str, genes_to_transcripts_path: str):
+    """Run tximport R script based on input exp_dir and the path of
+    genes_to_transcripts.txt."""
+
+    result = ComputationalResult()
+    cmd_tokens = [
+        "/usr/bin/Rscript", "--vanilla",
+        "/home/user/data_refinery_workers/processors/tximport.R",
+        "--exp_dir", exp_dir,
+        "--gene2txmap", genes_to_transcripts_path
+    ]
+    # TODO: Because this job is done at experiment level, it doesn't make
+    # sense to update job_context (which is at sample level right now).
+    # The DB model may have to be modified to associate this ComputationalResult
+    # record with the experiment (instead of samples in the experiment).
+    try:
+        result.time_start = timezone.now()
+        subprocess.run(cmd_tokens, stdout=subprocess.PIPE, stderr=subprocess.PIPE)
+        result.time_end = timezone.now()
+        result.command_executed = " ".join(cmd_tokens)
+        result.system_version = __version__
+        result.is_ccdl = True
+        result.save()
+    except Exception as e:
+        error_template = ("Encountered error in R code while running tximport.R"
+                          " pipeline during processing of {0}: {1}")
+        error_message = error_template.format(exp_dir, str(e))
+        logger.error(error_message, processor_job=job_context["job_id"])
```

Comment:
I think we should mark the job as a failure if this doesn't succeed. With the way this is right now, if this fails the only way we'll have any idea is if we see this error message in the logs.

Reply:
Sure. I can add a comment line about it.

Reply:
I didn't do it because even if `tximport` fails for whatever reason, the jobs that follows (salmontools, multiqc, etc) should still continue.

Reply:
I don't know that I agree that they should, however I'm not sure. This actually seems like a good question for @jaclyn-taroni: if `tximport` fails for some reason, would we prefer to have the rest of the pipeline continue or prevent that sample/experiment from being added to the compendium? It's not entirely clear to me how critical `tximport` is to have for a sample.

However, regardless of the answer to that question, I think we should probably fail the jobs at least for a while until we run into some errors that we can't find a workaround for, and then only let that kind of error through. The reason I think this is that as we run this on millions of samples, we're going to run into tons of edge cases which will cause failures. The way we're gonna debug these will be to look through our jobs/samples/experiments and figure out which didn't make it all the way through the system successfully and then track down what went wrong using the logs. We probably won't be going through every error message in the log files trying to figure out what happened, so if a job doesn't fail it won't make it onto our radar, and we won't realize that it failed for a fixable reason.

Reply:
> if tximport fails for some reason, would we prefer to have the rest of the pipeline continue or prevent that sample/experiment from being added to the compendium? It's not entirely clear to me how critical tximport is to have for a sample.

Hmm, this is a great question. I would say at this stage (Keytar Kurt) just fail the job, as right now (to my knowledge) the only things made available to the users are 1) count-scale output of `tximport` and 2) `multiqc` reports. In the future, we will probably want to allow users to obtain different outputs from the pipeline (e.g., transcript-level abundances from `salmon quant`, etc).

Reply:
Sounds good, thanks @dongbohu!